ext/ManoptManifoldsExt/manifold_functions.jl

```diff
@@ -58,3 +58,73 @@ function mid_point!(M::Sphere, y, p, q, x)
     y .= exp(M, p, 0.5 * X)
     return y
 end
+
+@doc raw"""
+    reflect(M, f, x; kwargs...)
+    reflect!(M, q, f, x; kwargs...)
+
+reflect the point `x` from the manifold `M` at the point `f(x)` of the
+function ``f: \mathcal M → \mathcal M``, given by
+
+````math
+    \operatorname{refl}_f(x) = \operatorname{refl}_{f(x)}(x),
+````
+
+Compute the result in `q`.
+
+see also [`reflect`](@ref reflect(M::AbstractManifold, p, x))`(M,p,x)`, to which the keywords are also passed to.
+"""
+reflect(M::AbstractManifold, pr::Function, x; kwargs...) = reflect(M, pr(x), x; kwargs...)
```

Comment:
Technically it's type piracy so maybe it would be better to introduce a new function for this functionality?

Reply:
Hm, complicated, because the reflection at the prox – which is what this is – does still bear the name reflection.

Yes I moved it here because Aqua said in the Manopt file it is type piracy, but in there? Hm, I am not sure.
But the worst is that I neither have a good name _and_ it would really be breaking to rename that, in the sense that it breaks all my Douglas-Rachford experiments.

Reply:
It's still piracy in the extension, Aqua just doesn't test for that yet: https://github.com/JuliaTesting/Aqua.jl/issues/160 . This is not a high priority thing but I'd consider renaming for the next breaking release.

Reply:
So you haven't fixed the problem, only swept it under the rug :sweat_smile:

Reply:
hm, yeah, maybe. 
I would maybe move this to Manifolds, that would resolve it,
just if one wants to be precise the “original” reflect is the manifold one while this is the Manopt one – so I am actually a bit unsure how to resolve this without piracy

Reply:
The only real solution I see here is, to move this to ManifoldsBase.jl as well.

Reply:
Reflecting at point like fine to me and we can move that to ManifoldsBase but the methods that dispatch on `Function` are questionable. What if someone wanted to represent points on a manifold by functions? We would have an ambiguity.

Reply:
The only answer I have to that is, that I have no clue about that. But for now we do not have that and I think representing points by functions would be a major rework anyways. In all 3 packages.

Reply:
I did that some time ago in FunManifolds.jl. That's a very niche application but I don't want to introduce new issues that could affect it if "let's make a new function with a different name" is all it takes to avoid it.

Reply:
But we could wrap that function in a subtype of ManifoldPoint?
The problem with points being functions is that _all_ which-leverl interfaces of Manopt.jl would not be able to handle them anyways. That is how I currently distinguish whether to create a random starting point, that a point is not a function.

My main problem with a new name is, that mathematically both have the same name here, so introducing a new name would confuse everyone reading the papers. So that takes all to confuse everyone, sure.

Reply:
That's fine in limited scenarios but note that the confusion may also work the other way: someone makes a functor object, doesn't derive from `Function` for whatever reason, and is surprised with a weird error. You can perhaps call the function variant `reflect_by`?

Reply:
So "something is function-like only if its type is `<:Function` is also a flawed assumption.

Reply:
Sure, but in some places I did not yet see any other way for the “convenience” layer to distinguish between a Hessian being provided (and not needed to be approximated) or a point (and no random is needed). And maybe this is not a good design, but I am not clever enough to come up with a better one that is simple to use.

As I said, I do not have a good idea for a name. the `_by` would sound strange, for both a point and a function one would say you reflect _at_, but then this does not help, since it would refer to both.

Reply:
> So "something is function-like only if its type is <:Function is also a flawed assumption.

Yes, sure.

Reply:
But I also only do that on the high-level interfaces `solver(M, f, ...)`not on problem/state.

The idea is (also maybe not good) that as long as you are in the simple regime, the high level stuff should be easy to use.

Reply:
Dispatching on functions is, as I said, fine in limited scenarios. So high-level solver interface doing it is not a big issue. But I don't want to propagate that design to ManifoldsBase.jl, especially when the alternative is just splitting it into two functions.

Reply:
I will think about a name. But me and thinking about names is something random between tomorrow and next year :/

Reply:
So let's keep `reflect` in Manopt until you figure it out :slightly_smiling_face:

Reply:
Yes, I think that for now is the best. And I for now also feel that the piracy here is also only a mild one, I define a function based on a ManifoldsBase type, that in principle I “own” as well, but sure with `reflect!` I pirate then a bit since the function also comes from LinearAlgreba.

But yes I will keep that in mind and think about it to hopefully find a good alternative, maybe something keyword based `at_point=` and `at_function` or such (just the function name `reflect_at_point` seems to long for me again).

Reply:
Sure, it can also be handled by keywords.